### ISSUE TITLE
User Status Management Updates

### ISSUE THREAD

**Issue by heyrandhir** [Real-Dev-Squad/website-backend#1184]
**Parent Ticket:** https://github.com/Real-Dev-Squad/todo-action-items/issues/155

**Summary:**
Update user status management on "My Site" to remove the ability for users to manually mark themselves as Idle or Active. Implement a task-based status change logic in the backend as per the workflow described.

**Details:**
1. Currently, users on "My Site" have the ability to manually mark themselves as Idle, Active, or OOO. However, this functionality will be removed from the site.
2. Users will only be able to mark themselves as OOO from "My Site." All other status changes (Idle and Active) will be determined based on task assignments and task completion.
3. Implement the following workflow for user status changes:
   - When a task is assigned to a user, their status should be automatically set as Active if their current status is not already Active (i.e., Idle or Onboarding).
   - If a user is marked as OOO, their future status should be set as Active, allowing them to resume their tasks upon returning.
   - When a user completes a task, the system should check if there are any existing assigned or in-progress tasks for that user.
     - If there are assigned or in-progress tasks, the user's status should remain Active.
     - If there are no assigned or in-progress tasks, the system should check if the user is OOO.
       - If the user is not OOO, their status should be marked as Idle.
       - If the user is OOO, their future status should be marked as OOO.
4. Update the backend logic to handle the task-based status changes as per the workflow mentioned above.

**UML Sequence Diagram:**

The UML sequence diagram depicting the updated user status management logic has been attached for reference. It illustrates the revised flow of actions and interactions between the user and the system components.

1. UML Sequence Diagram for Status update from My Site 
![image](https://github.com/Real-Dev-Squad/website-backend/assets/97341921/d0fa3f59-6ea9-493e-a787-42993f29163d)

2. UML Sequence Diagram for User Status update when New Task is Created or Updated (Website Backend)
![image](https://github.com/Real-Dev-Squad/website-backend/assets/97341921/011c3418-32c7-4aa2-ba8f-fd60b3f2914d)

3. UML Sequence Diagram for User Status update post Task Completion (Website Backend)
![image](https://github.com/Real-Dev-Squad/website-backend/assets/97341921/04dd0e75-39ff-4d6e-b813-541b2bafde66)

**Comment by heyrandhir** [comment]
![image](https://github.com/Real-Dev-Squad/website-backend/assets/97341921/ade1534d-22b2-408b-bfe3-df1c7c64897c)

**Comment by heyrandhir** [comment]
Hi @ankushdharkar 

If we remove the option to mark oneself as "idle" or "active" from our my site, how would a user mark themselves as "active" again if they have set their status as "out of office" (OOO) but then decide to become available before their previously designated OOO period has ended. Even if a super user create new task and assign it to them during this time, their status will still show as OOO and their status would not be able to change to active on creation of new task.

**Comment by heyrandhir** [comment]
> Hi @ankushdharkar
> 
> If we remove the option to mark oneself as "idle" or "active" from our my site, how would a user mark themselves as "active" again if they have set their status as "out of office" (OOO) but then decide to become available before their previously designated OOO period has ended. Even if a super user create new task and assign it to them during this time, their status will still show as OOO and their status would not be able to change to active on creation of new task.

During the standup meeting, @ankushdharkar mentioned that for now we will implement a new button on My Site that will give the users the ability to cancel  Out of Office.

**Comment by prakashchoudhary07** [comment]
Nice 🚀 
Looks good

**Comment by heyrandhir** [comment]
There are 3 main parts to this proposed change:   

1. Update the status when a task is completed. 
2. Change the status when a task is assigned to or updated by the super user.
3. Add the ability for a user to cancel their OOO before their OOO period has ended.
4. Allow Users to mark themselves as "Out of Office" (OOO) only. The "idle" and "active" statuses would no longer be user-selectable.

**Comment by heyrandhir** [comment]
Here is the updated ETA for the 4 proposed changes:

1. Update status on task completion.
2. Change status on task assign/update.
3. Add ability to cancel OOO.
4. Allow only OOO status.

For each change, I've allocated:

API Contract - 0.5 day 
Testing - 1 day
Models/Controllers - 1 day   
Refactoring/PR changes - 2 day
Buffer - 0.5 day
Total - 4- 5 days
 
I can refine the timelines as we proceed further and I get a better sense of the actual effort required.
